File: String/LeetCode3.go
package string

func lengthOfLongestSubstring(s string) int {
	l, r, res := 0, 0, 0
	cnt := make(map[byte]int, len(s))
	for r < len(s) {
		c1 := s[r]
		cnt[c1]++
		r++
		//如果cnt中出现重复字符
		//开始移动l来缩小窗口
		for cnt[c1] > 1 {
			c2 := s[l]
			cnt[c2]--
			l++
		}
		res = max(res, r-l)
	}
	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: Array/twopointer/LeetCode3.cc
#include<iostream>
#include<vector>
#include<unordered_set>
#include<algorithm>
using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        //使用哈希表记录每个字符是否出现过
        unordered_set<char> sumSet;
        int n = s.size();
        int rk = -1, ans = 0;
        for(int i = 0; i < n; i++){
            if(i != 0){
                sumSet.erase(s[i-1]);
            }

            while(rk+1 < n && !sumSet.count(s[rk+1])){
                sumSet.insert(s[rk+1]);
                ++rk;
            }
            ans = max(ans,rk-i+1);
        }
        return ans;
    }
};